list.lisp

```diff
@@ -0,0 +1,83 @@
+;;;; -*- Mode: Lisp; indent-tabs-mode: nil -*-
+
+(in-package :split-sequence)
+
+(declaim (inline
+          split-list split-list-if split-list-if-not
+          split-list-from-end split-list-from-start split-list-internal))
+
+(declaim (ftype (function (&rest t) (values list unsigned-byte))
+                split-list split-list-if split-list-if-not))
+
+(declaim (ftype (function (function list unsigned-byte (or null unsigned-byte) (or null unsigned-byte)
+                                    boolean function)
+                          (values list unsigned-byte))
+                split-list-from-start split-list-from-end split-list-internal))
+
+(defun split-list-internal (predicate list start end count remove-empty-subseqs key)
```

Comment:
This is kind of a long function.  With some `loop`s and `flet`s it could be significantly shorter, and maybe a little clearer:

```lisp
(defun split-list-internal (predicate list start end count remove-empty-subseqs key)
  (let ((result '())
        (index start)
        (count count)
        (list (nthcdr start list)))
    (flet ((push-chunk (chunk)
             (unless (and remove-empty-subseqs (null chunk))
               (push chunk result)
               (when count (decf count))))
           (gather-chunk ()
             (loop :for value = (pop list)
                   :while (< index end)
                   :do (incf index)
                   :until (funcall predicate (funcall key value))
                   :collect value)))
      (loop :while (and list
                        (< index end)
                        (or (null count) (plusp count)))
            :do (push-chunk (gather-chunk))
            :finally (return (values (if (null result)
                                       (list nil)
                                       (nreverse result))
                                     index))))))
```

`(loop :collect)` handles all the consing-at-the-end stuff so you don't have to keep track of the last cons yourself.

I did a little benchmarking and there's pretty much no difference performance-wise, at least on my machine.

Reply:
Test:
```
CL-USER> (time (split-sequence:split-sequence-if (lambda (x) (= 0 (mod x 5)))
                                                 (alexandria:iota 20000000)))
```

My version:
```
Evaluation took:
  1.224 seconds of real time
  1.224159 seconds of total run time (1.076714 user, 0.147445 system)
  [ Run times consist of 0.320 seconds GC time, and 0.905 seconds non-GC time. ]
  100.00% CPU
  4,029,522,115 processor cycles
  703,987,712 bytes consed
```

Your version:

```
Evaluation took:
  1.435 seconds of real time
  1.435424 seconds of total run time (1.203668 user, 0.231756 system)
  [ Run times consist of 0.585 seconds GC time, and 0.851 seconds non-GC time. ]
  100.00% CPU
  4,723,886,276 processor cycles
  704,120,592 bytes consed
```

The consing difference is negligible but run time increases by about 17%.

Given the huge difference in readability and the fact that we nonetheless gain a lot of speed compared to #9, I think it is good to make this change.

Reply:
Your function fails the test suite though.

Example failures:

```
## SPLIT-SEQUENCE.6.LIST: 
(VALUES OUTPUT INDEX)
 evaluated to 
((NIL (#\b #\r) (#\c) (#\d) (#\b #\r)) 11)
 which is not 
EQUALP
 to 
((NIL (#\b #\r) (#\c) (#\d) (#\b #\r) NIL) 11)
```
```
## SPLIT-SEQUENCE.10.LIST: 
(VALUES OUTPUT INDEX)
 evaluated to 
((NIL) 9)
 which is not 
EQUALP
 to 
(NIL 8)
```

Your code skips an empty sequence when SPLIT-SEQUENCE outputs one, and outputs an empty sequence when SPLIT-SEQUENCE skips one.

These edge cases is exactly the kind of behavior that led me to dropping LOOP and implementing the cons-based monstrosity you see.

Reply:
Fixed in 3fb1ab8